cpp/src/copying/slice.cpp

```diff
@@ -36,7 +36,22 @@ std::vector<column_view> slice(column_view const& input,
 
   if (indices.empty()) return {};
 
-  auto null_counts = cudf::detail::segmented_count_unset_bits(input.null_mask(), indices, stream);
+  auto null_counts = [&]() {
+    // need to shift incoming indices by the column offset to generate the correct bit ranges
+    // to count
+    if (input.offset() > 0) {
+      std::vector<size_type> shifted_indices;
+      shifted_indices.reserve(indices.size());
+      std::transform(indices.begin(),
+                     indices.end(),
+                     std::back_inserter(shifted_indices),
+                     [offset = input.offset()](size_type index) { return index + offset; });
```

Comment:
`slice()` is supposed to be a relatively free operation (ignoring the bit counting), so materializing the shifted indices is a bummer.

I think it would be fairly easy to make `segmented_count_unset_bits` take an iterator for the segment offsets so you could instead pass a transform iterator in that adds the offset.

Reply:
How about just invalidating the null count so the slice function will never do such counting? We don't know if the null count will be used later or not. If it is used later, it will be recomputed.

Reply:
We used to do it that way, but we found that it was a better idea to compute all the null counts at once. Otherwise if you have many slices where you need the null count, you end up launching many kernels. This came up in the context of contiguous_split.

Reply:
Note: slice.cpp will have to become slice.cu to do this.

Reply:
Done.  Couple things of note:  A number of functions/kernels got moved from `null_mask.cu` to `null_mask.cuh`. Both `segmented_count_unset_bits` and `segmented_count_set_bits` have been refactored to take iterators, since `unset` is implemented through `set`.  I had to change a device lambda in `segmented_count_unset_bits` into a functor (`word_num_set_bits_functor`) due to extended device lambda restrictions causing the compiler to complain about the iterators now being passed in.